src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs

```diff
@@ -410,7 +507,7 @@ protected override void VisitSwitchSection(BoundSwitchSection node, bool isLastS
                                         learnFromNonNullTest(inputSlot, ref this.StateWhenTrue);
                                     }
                                     gotoNode(p.WhenTrue, this.StateWhenTrue, nodeBelievedReachable);
-                                    gotoNode(p.WhenFalse, this.StateWhenFalse, nodeBelievedReachable & inputState.MayBeNull());
+                                    gotoNode(p.WhenFalse, this.StateWhenFalse, nodeBelievedReachable & inputMaybeNull);
```

Comment:
It looks like this introduces the effect where, if an input has not-null flow state, we assume that dag nodes which rely on the input actually being `null` are unreachable. But in general when the user null tests an input, we sort of do the opposite and say: oh, now we know that it actually could be null. Which scenarios are affected by this?

Reply:
Without this check, only a couple of tests break.
Here's one:
```
    int M1(string s1, string s2)
    {
        return (s1, s2) switch {
            (string x, string y) => x.Length + y.Length
            };
    }
```
That scenario gets a new warning:
```
 // (6,25): warning CS8655: The switch expression does not handle some null inputs (it is not exhaustive). For example, the pattern '(null, _)' is not covered.
                    //         return (s1, s2) switch {
                    Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustiveForNull, "switch").WithArguments("(null, _)").WithLocation(6, 25),
```
So this isn't so much "treating pure null tests seriously" as "consider branches unreachable when there is an implicit null-test but the input is not-null".

Pure null tests are handled in the `BoundDagExplicitNullTest` arm below. Those will change an input from not-null to maybe-null.

---
In reply to: [574012059](https://github.com/dotnet/roslyn/pull/51001#discussion_r574012059) [](ancestors = 574012059)